src/mpid/ch4/netmod/ofi/ofi_impl.h

```diff
@@ -33,8 +33,28 @@ ATTRIBUTE((unused));
 #define MPIDI_OFI_COMM(comm)     ((comm)->dev.ch4.netmod.ofi)
 #define MPIDI_OFI_COMM_TO_INDEX(comm,rank) \
     MPIDIU_comm_rank_to_pid(comm, rank, NULL, NULL)
-#define MPIDI_OFI_TO_PHYS(avtid, lpid, _nic) \
-    MPIDI_OFI_AV(&MPIDIU_get_av((avtid), (lpid))).dest[_nic][0]
+
+#define MPIDI_OFI_AV_ROOT_ADDR(av)      MPIDI_OFI_AV(av).root_dest
+
+#ifdef MPIDI_OFI_VNI_USE_DOMAIN
+#define MPIDI_OFI_AV_ADDR_ROOT(av) \
+    MPIDI_OFI_AV(av).root_dest
+#define MPIDI_OFI_AV_ADDR_OFFSET(av, vci, nic) \
+    (MPIDI_OFI_AV(av).all_dest[(vci)*MPIDI_OFI_global.num_nics+(nic)] + MPIDI_OFI_AV(av).root_offset)
+#define MPIDI_OFI_AV_ADDR_NO_OFFSET(av, vci, nic) \
+    MPIDI_OFI_AV(av).all_dest[(vci)*MPIDI_OFI_global.num_nics+(nic)]
+#else /* scalable endpoints - all vci share the same addr */
+#define MPIDI_OFI_AV_ADDR_ROOT(av) \
+    MPIDI_OFI_AV(av).root_dest
+#define MPIDI_OFI_AV_ADDR_OFFSET(av, vci, nic) \
+    (MPIDI_OFI_AV(av).all_dest[nic] + MPIDI_OFI_AV(av).root_offset)
+#define MPIDI_OFI_AV_ADDR_NO_OFFSET(av, vci, nic) \
+    MPIDI_OFI_AV(av).all_dest[nic]
+#endif
+#define MPIDI_OFI_AV_ADDR(av, local_vci, local_nic, vci, nic) \
+    ((local_vci==0 && local_nic==0) ? \
+        ((vci == 0 && nic == 0) ? MPIDI_OFI_AV_ADDR_ROOT(av) : MPIDI_OFI_AV_ADDR_OFFSET(av, vci, nic)) : \
+        MPIDI_OFI_AV_ADDR_NO_OFFSET(av, vci, nic))
```

Comment:
can we use if/else instead of double ternary here? hard to parse what is going on.

Reply:
`if/else` is not an expression. I can try make it into an inline function instead.

Reply:
I can't make it into inline function because it is used as a left value (that being assigned to). However, it is only used in `ofi_vci.c`, so let me move it into the file and then see if I can make the code more readable.

Reply:
I cleaned it up in the last commits. It is more difficult to fix the commits as it first appears since it is getting refactored.